qlib/contrib/model/pytorch_gru.py

```diff
@@ -0,0 +1,362 @@
+            cur_loss.backward()
+            self.train_optimizer.step()
+            loss.update(cur_loss.item())
+
+            # validation
+            train_loss += loss.val
+            # print(loss.val)
+            if step and step % self.eval_steps == 0:
+                stop_steps += 1
+                train_loss /= self.eval_steps
+
+                with torch.no_grad():
+                    self.gru_model.eval()
+                    loss_val = AverageMeter()
+
+                    # forward
+                    preds = self.gru_model(x_val_auto)
+                    cur_loss_val = self.get_loss(preds, y_val_auto, self.loss_type)
+                    loss_val.update(cur_loss_val.item())
+
+                if verbose:
+                    self.logger.info(
+                        "[Epoch {}]: train_loss {:.6f}, valid_loss {:.6f}".format(step, train_loss, loss_val.val)
+                    )
+                evals_result["train"].append(train_loss)
+                evals_result["valid"].append(loss_val.val)
+                if loss_val.val < best_loss:
+                    if verbose:
+                        self.logger.info(
+                            "\tvalid loss update from {:.6f} to {:.6f}, save checkpoint.".format(
+                                best_loss, loss_val.val
+                            )
+                        )
+                    best_loss = loss_val.val
+                    stop_steps = 0
+                    torch.save(self.gru_model.state_dict(), save_path)
+                train_loss = 0
+                # update learning rate
+                self.scheduler.step(cur_loss_val)
+
+        # restore the optimal parameters after training ??
+        # self.gru_model.load_state_dict(torch.load(save_path))
+        if self.use_gpu:
+            torch.cuda.empty_cache()
+
+    def get_loss(self, pred, target, loss_type):
+        if loss_type == "mse":
+            sqr_loss = (pred - target)**2
+            loss = sqr_loss.mean()
+            return loss
+        elif loss_type == "binary":
+            loss = nn.BCELoss()
+            return loss(pred, target)
+        else:
+            raise NotImplementedError("loss {} is not supported!".format(loss_type))
+
+    def predict(self, dataset):
+        if not self._fitted:
+            raise ValueError("model is not fitted yet!")
+
+        x_test = dataset.prepare("test", col_set="feature")
+        x_test = x_test.fillna(0)
+        index = x_test.index
+        x_test = torch.from_numpy(x_test.values).float()
+
+        if self.use_gpu:
+            x_test = x_test.cuda()
+        self.gru_model.eval()
+
+        with torch.no_grad():
+            if self.use_gpu:
+                preds = self.gru_model(x_test).detach().cpu().numpy()
+            else:
+                preds = self.gru_model(x_test).detach().numpy()
+        return pd.Series(preds, index=index)
+
+    def save(self, filename, **kwargs):
```

Comment:
Models  are subclass of [Serializable](https://github.com/you-n-g/qlib/blob/main/qlib/utils/serial.py#L8)
We  plan to use a pickle-friendly interface to save and load models(including handler, processor . etc.)

If there are something special to save and load data, please override  `__setstate__` , `__getstate__`,   methods

Reply:
If you have any inconvenience using such interface, please discuss with about the design.

Reply:
ok, I'll delete the save and load function.